iothub/device/src/RetryPolicies/RetryHandler.cs

```diff
@@ -0,0 +1,152 @@
+        /// </returns>
+        internal async Task<T> RunWithRetryAsync<T>(Func<Task<T>> taskFunc, CancellationToken cancellationToken = default)
+        {
+            return await RunWithRetryAsync(taskFunc, (ex) => false, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Repetitively executes the specified asynchronous task with a specific return value while it satisfies the current retry policy.
+        /// </summary>
+        /// <param name="taskFunc">A function that returns a started task (also known as "hot" task).</param>
+        /// <param name="transientErrorCheck">Additional check for transient exceptions.</param>
+        /// <param name="cancellationToken">The token used to cancel the retry operation. This token does not cancel the execution
+        /// of the asynchronous task.</param>
+        /// <returns>
+        /// Returns a task that will run to completion if the original task completes successfully (either the
+        /// first time or after retrying transient failures). If the task fails with a non-transient error or
+        /// the retry limit is reached, the returned task will transition to a faulted state and the exception must be observed.
+        /// </returns>
+        internal async Task<T> RunWithRetryAsync<T>(Func<Task<T>> taskFunc, Func<Exception, bool> transientErrorCheck, CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            Debug.Assert(taskFunc != null);
+            Debug.Assert(transientErrorCheck != null);
+
+            uint retryCount = 0;
+            TimeSpan retryDelay;
+            Exception lastException;
+
+            do
+            {
+                try
+                {
+                    return await taskFunc().ConfigureAwait(false);
+                }
+                catch (Exception ex) when (!cancellationToken.IsCancellationRequested
+                    && (transientErrorCheck.Invoke(ex)
+                        || _retryPolicy.ShouldRetry(++retryCount, ex, out retryDelay)))
+                {
+                    lastException = ex;
+                }
+
+                if (retryDelay < TimeSpan.Zero)
+                {
+                    retryDelay = TimeSpan.Zero;
+                }
+
+                // If we expect to wait until retry, calculate the remaining wait time, considering how much time the operation already
+                // took, and the minimum delay.
+                if (retryDelay > TimeSpan.Zero
+                    && !cancellationToken.IsCancellationRequested)
+                {
+                    // Don't pass in the cancellation token, because we'll handle that
+                    // condition specially in the catch blocks above.
+                    await Task.Delay(retryDelay, CancellationToken.None).ConfigureAwait(false);
```

Comment:
Vastly simplified this. Got rid of a min wait and a timer to see how long the operation took and subtracting that from the next delay.